### PR TITLE
Move new default initializers into separate directory

### LINKED ISSUE #25181: Move new default initializers into their own directory
We have a number of new initializers that are generated for new apps to prevent old apps from hitting breaking changes. They are:

```
active_record_belongs_to_required_by_default.rb
callback_terminator.rb
to_time_preserves_timezone.rb
per_form_csrf_tokens.rb
cookies_serializer.rb
request_forgery_protection.rb
session_store.rb
```

These are not files that anyone is likely to edit by hand on a new app, so they shouldn't litter the top initializers directory. Let's move all these into `config/initializers/new_framework_defaults/*`.

### PR DESCRIPTION
**Pull request by prathamesh-sonpatki**
r? @dhh 
- Fixes #25181

### CODE REVIEW

File: railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults/ssl_options.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Be sure to restart your server when you modify this file.
+
+# Configure SSL options to enable HSTS with subdomains. This is a new
+# Rails 5.0 default, so it is introduced as a configuration option to ensure
+# that apps made on earlier versions of Rails are not affected when upgrading.
+Rails.application.config.ssl_options = { hsts: { subdomains: true } }
```
Comment (prathamesh-sonpatki) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults/ssl_options.rb]: This is also new Rails 5 default  so moved it too.

**Comment by prathamesh-sonpatki** [comment]
@dhh Moved `per_form_csrf_tokens.rb` and `request_forgery_protection.rb` also into new directory.

**Comment by dhh** [comment]
👍. Have you verified that the boot sequence loads config/intializers/**/* and not just config/initializers/*?

**Comment by prathamesh-sonpatki** [comment]
Yes. I created a new app and tested the `belongs_to` default by trying to create a post without corresponding user and it failed.

**Comment by dhh** [comment]
Sweet. Will merge when the build returns, if there are no novel failures.

Thanks for working on both of these so quickly!! I'm going through what a new app generated with both API mode and without looks like. config/initializers was definitely turning junky. This cleans it up nicely.

**Comment by prathamesh-sonpatki** [comment]
I tested app generator tests and they are passing. Hopefully that will be enough :smile:

**Comment by prathamesh-sonpatki** [comment]
I was wrong. There are some failures. Looking into those.

**Comment by prathamesh-sonpatki** [comment]
@dhh Build is green now.

**Comment by dhh** [comment]
Thinking about this some more, I think we should actually just collapse all these new defaults into 1 file: config/initializers/new_framework_defaults.rb. What we have now is still an improvement, but the whole point of these new defaults is that it's just there for backwards compatibility. Not for people to necessarily tinker with on new apps. So I think a single file shows that more proportionally. Thoughts @prathamesh-sonpatki ?

**Comment by prathamesh-sonpatki** [comment]
@dhh Good idea. Currently we do some housekeeping about the new initializers -

https://github.com/prathamesh-sonpatki/rails/blob/master/railties/lib/rails/generators/rails/app/app_generator.rb#L345

https://github.com/prathamesh-sonpatki/rails/blob/master/railties/lib/rails/generators/rails/app/app_generator.rb#L363-L364

https://github.com/prathamesh-sonpatki/rails/blob/master/railties/lib/rails/generators/rails/app/app_generator.rb#L92-L97

and so on. 

I think we can move all this logic `config/initializers/new_framework_defaults.rb` and generate it based on `options[:api]` or `options[:skip_active_record]` and so on.

Also for older apps, we just have to make sure that this file does not get created instead of all the logic in https://github.com/prathamesh-sonpatki/rails/blob/master/railties/lib/rails/generators/rails/app/app_generator.rb#L104-L117.

I can give it a go :smile:

**Comment by dhh** [comment]
Awesome, thanks!

**Comment by kaspth** [comment]
Another thing to consider: once we're past the migration release, should we remove past configs in the file when running `bin/rails app:update`?

E.g. if the purpose of the cookie serializer is to use JSON going forward, but the option to use Marshal when upgrading, should we gsub that config out on the Rails version after JSON has become the default?

**Comment by prathamesh-sonpatki** [comment]
@kaspth Once we switch to something as default in the code (not from initializers) then we can remove the `gsub` housekeeping. But before that we can give users chance to set their own config. 

For eg. for cookies serializer, `JSON` can be default for new apps and initializer will be removed for new apps. 

Two scenarios for old apps -- either the config is set(via an initializer from 4.1) or not set. 

If not set, then we set it to `hybrid` and print out a deprecation warning that in next rails release, unset config will be treated as `JSON`.  So we no longer have to do `gsub` for old apps because we are taking care of it in code. 

If the config is set, we just use it.

In next release, `JSON` becomes the default from the code. Also we no longer `gsub` for old apps.

This is just one specific example related to cookies_serializer but I think can be adapted to others as needed.

**Pull request merged**